# Ebay Car Sales Data #

We will work with a dataset of used cars from eBay Klieinanzeigen, a classifieds section of the German eBay website. The aim of the project is to clean the data and analyze the included used car listings.

The data dictionary provided with data as follows:
* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered
* `gearbox` - the transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthofRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - if the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

## Importing and Reading Data ##

In [137]:
import pandas as pd
import numpy as np

In [138]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [139]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [140]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [141]:
#Check the missing values 
missing_values = autos.isnull().sum()
missing_values.sort_values(ascending = False)

#Find percentage of missing values 
percentage_missing = round((missing_values/len(autos)) * 100)
percentage_missing.sort_values(ascending = False)

notRepairedDamage      20.0
vehicleType            10.0
fuelType                9.0
model                   6.0
gearbox                 5.0
lastSeen                0.0
yearOfRegistration      0.0
name                    0.0
seller                  0.0
offerType               0.0
price                   0.0
abtest                  0.0
powerPS                 0.0
postalCode              0.0
odometer                0.0
monthOfRegistration     0.0
brand                   0.0
dateCreated             0.0
nrOfPictures            0.0
dateCrawled             0.0
dtype: float64

In [142]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**Observations** 
- Total 20 columns, 5000 rows 
- Data types are mostly strings, but few integers 
- The columns ```Vehicle Type```, ```Gear Box, Model```, ```Fuel Type```, ```Not Repaired Damage``` have null values (no more than ~20% null values)
- Column names use camelcase instead of Python's preferred snakecase

In [143]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [144]:
#Change the column names to appropriate snakecase letters
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type',
                 'price', 'abtest', 'vehicle_type', 'registeration_year',
                 'gearbox', 'power_ps', 'model', 'odometer',
                 'registeration_month', 'fuel_type', 'brand',
                 'unrepaired_damage', 'ad_created', 'nr_of_pictures',
                 'postal_code', 'last_seen']

In [145]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registeration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registeration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [146]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registeration_year  \
0  privat    Angebot  $5,000  control          bus                2004   
1  privat    Angebot  $8,500  control    limousine                1997   
2  privat    Angebot  $8,990     test    limousine                2009   
3  privat    Angebot  $4,350  control   kleinwagen                2007   
4  privat    Angebot  $1,350     test        kombi                2003   

     gearbox  power_ps   model   odometer  registeration_month fuel_type  \
0    manuell       158  andere  150,000km                    3       lpg   
1  automatik       286     7er  150,000km                    6    benzin   
2    manuell       102    golf   70,000km                    7    benzin   
3  automatik        71  fortwo   70,000km                    6    benzin   
4    manuell         0   focus  150,000km                    7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

- I had to change the column names from camelcase to snakecase so Python can read them. 

## Initial Exploration and Cleaning ##
### What to look out for ###
- The text columns where most of the values are the same, those need to be dropped. They do not provide us with useful information for analysis. 
- Numeric data stored as text needs to be converted to numbers and cleaned  


In [147]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registeration_year  gearbox      power_ps  model  \
count         44905        50000.000000    47320  50000.000000  47242   
unique            8                 NaN        2           NaN    245   
top       limousine                 NaN  manuell           NaN   golf   
freq          12859                 NaN    36993           NaN   4024   
mean            NaN         2005.073280      NaN    116.355920    NaN   
std             NaN          105.712813      NaN    209.216627    NaN   
min             NaN         1000.000000      NaN      0.000000    NaN   
25%             NaN         1999.000000      NaN     70.000000    NaN   
50%             NaN         2003.000000      NaN    105.000000    NaN   
75%             NaN         2008.000000      NaN    150.000000    NaN   
max             NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  registeration_month fuel_type       brand  \
count       50000         50000.000000     45518       50000   
unique         13                  NaN         7          40   
top     150,000km                  NaN    benzin  volkswagen   
freq        32424                  NaN     30107       10687   
mean          NaN             5.723360       NaN         NaN   
std           NaN             3.711984       NaN         NaN   
min           NaN             0.000000       NaN         NaN   
25%           NaN             3.000000       NaN         NaN   
50%           NaN             6.000000       NaN         NaN   
75%           NaN             9.000000       NaN         NaN   
max           NaN            12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [148]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,200      639
           ... 
$29,777       1
$13,888       1
$34,750       1
$3,725        1
$8,777        1
Name: price, Length: 2357, dtype: int64

In [149]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [150]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [151]:
reg_year = autos['registeration_year'].unique()
np.sort(reg_year)

array([1000, 1001, 1111, 1500, 1800, 1910, 1927, 1929, 1931, 1934, 1937,
       1938, 1939, 1941, 1943, 1948, 1950, 1951, 1952, 1953, 1954, 1955,
       1956, 1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966,
       1967, 1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977,
       1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988,
       1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
       2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
       2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2800, 4100,
       4500, 4800, 5000, 5911, 6200, 8888, 9000, 9996, 9999])

In [152]:
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

## Clean Columns ##
### Summary of Descriptive Stats ###

- ```price``` and ```odometer``` columns need to be converted to numeric types from text types 
- ```registeration year``` contains invalid years because it ranges from 1000 to 9999 years
- ```nr_of_pictures``` can be dropped because it only contains 0 as its value 
- ```seller``` and ```offer types``` have almost the same values 

In [153]:
#Remove any non-numeric characters in the data
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')

autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')

#Convert columns to numeric dtype
autos['price'] = autos['price'].astype(float)
autos['odometer'] = autos['odometer'].astype(float)

In [154]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  registeration_year  \
0      privat    Angebot   5000.0  control          bus                2004   
1      privat    Angebot   8500.0  control    limousine                1997   
2      privat    Angebot   8990.0     test    limousine                2009   
3      privat    Angebot   4350.0  control   kleinwagen                2007   
4      privat    Angebot   1350.0     test        kombi                2003   
...       ...        ...      ...      ...          ...                 ...   
49995  privat    Angebot  24900.0  control    limousine                2011   
49996  privat    Angebot   1980.0  control       cabrio                1996   
49997  privat    Angebot  13200.0     test       cabrio                2014   
49998  privat    Angebot  22900.0  control        kombi                2013   
49999  privat    Angebot   1250.0  control    limousine                1996   

         gearbox  power_ps   model  odometer  registeration_month fuel_type  \
0        manuell       158  andere  150000.0                    3       lpg   
1      automatik       286     7er  150000.0                    6    benzin   
2        manuell       102    golf   70000.0                    7    benzin   
3      automatik        71  fortwo   70000.0                    6    benzin   
4        manuell         0   focus  150000.0                    7    benzin   
...          ...       ...     ...       ...                  ...       ...   
49995  automatik       239      q5  100000.0                    1    diesel   
49996    manuell        75   astra  150000.0                    5    benzin   
49997  automatik        69     500    5000.0                   11    benzin   
49998    manuell       150      a3   40000.0                   11    diesel   
49999    manuell       101  vectra  150000.0                    1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  

In [155]:
#Rename column name - odometer to odometer_km 
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [156]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registeration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registeration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [157]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [158]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

### Exploring the Odometer and Price Columns ###
- ```odometer_km``` and ```price``` values do not look right
- These two columns are critical for our analysis  

In [159]:
autos['odometer_km'].unique().shape

(13,)

In [160]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [161]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

Based on our analysis, there seems to be no major outliers in the ```odometer_km``` column

In [162]:
autos['price'].unique().shape

(2357,)

In [163]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [164]:
autos['price'].value_counts().sort_index(ascending=False)

99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price, Length: 2357, dtype: int64

In [165]:
autos['price'].min(), autos['price'].max(), autos['price'].median(), autos['price'].mean()

(0.0, 99999999.0, 2950.0, 9840.04376)

In the ```price``` column, we need to remove the numbers between 0 and 1, and any numbers above 10 million, because they are all outliers. 
We should use boolean filter and filter index to remove these outliers 

In [166]:
#Find all price values between $0 and $5
price_0to5_bool = autos["price"].between(0, 5)
price_0to5_bool = autos[price_0to5_bool]
price_0to5_bool.index.name = 'index_0to5'

#Remove the price values between $0 and $5 - they are outliers 
autos.drop(price_0to5_bool.index, inplace=True)

In [167]:
autos['price'].describe()

count    4.841700e+04
mean     1.016176e+04
std      4.889028e+05
min      8.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.499000e+03
max      1.000000e+08
Name: price, dtype: float64

There are 1583 data values with ```price``` ranging between 0 and 5. 

In [168]:
#Find all prices values above $10 million
price_above10mill = autos['price'] > 10000000
len(autos[price_above10mill])

#Remove the price values above $10 million  - they are outliers 
autos.drop(autos[price_above10mill].index, inplace=True)

There are 7 data values with ```price``` above 10 million 

In [169]:
autos['price'].min(), autos['price'].max(), autos['price'].median(), autos['price'].mean()

(8.0, 10000000.0, 3000.0, 6309.038132617228)

Evaluate date columns to understand the date range the data covers 
- `date_crawled`, `last_seen`, `ad_created` columns are represented as strings, we need to convert the data into numerical representation to understand it quantitatively. 
    - `series.value_counts()` for distribution 
    - `series.str[:10]` to extract just the date values 
- `registeration_month` and `registeration_year` are numeric values, we can use `series.describe()` to understand the distribution 

In [170]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [171]:
date_crawled_day = autos['date_crawled'].str[:10]

In [172]:
#Calculate date_crawled frequency in percentages 
date_crawled_day.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025367
2016-03-06    0.014067
2016-03-07    0.036046
2016-03-08    0.033258
2016-03-09    0.033030
2016-03-10    0.032204
2016-03-11    0.032597
2016-03-12    0.036955
2016-03-13    0.015658
2016-03-14    0.036625
2016-03-15    0.034270
2016-03-16    0.029498
2016-03-17    0.031646
2016-03-18    0.012911
2016-03-19    0.034745
2016-03-20    0.037802
2016-03-21    0.037327
2016-03-22    0.032948
2016-03-23    0.032266
2016-03-24    0.029395
2016-03-25    0.031564
2016-03-26    0.032245
2016-03-27    0.031109
2016-03-28    0.034848
2016-03-29    0.034146
2016-03-30    0.033733
2016-03-31    0.031812
2016-04-01    0.033733
2016-04-02    0.035489
2016-04-03    0.038566
2016-04-04    0.036521
2016-04-05    0.013076
2016-04-06    0.003161
2016-04-07    0.001384
Name: date_crawled, dtype: float64

`date_crawled` range from 05-Mar-2016 to 07-Apr-2016 frequency percentage is less than 4%

In [173]:
#Calculate ad_created frequency in percentages 
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038814
2016-04-04    0.036893
2016-04-05    0.011795
2016-04-06    0.003243
2016-04-07    0.001239
Name: ad_created, Length: 76, dtype: float64

`ad_created` range from 11-Jun-2015 to 07-Apr-2016 frequency percentage is less than 4%

In [174]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001074
2016-03-06    0.004338
2016-03-07    0.005412
2016-03-08    0.007354
2016-03-09    0.009626
2016-03-10    0.010618
2016-03-11    0.012373
2016-03-12    0.023776
2016-03-13    0.008862
2016-03-14    0.012621
2016-03-15    0.015864
2016-03-16    0.016443
2016-03-17    0.028073
2016-03-18    0.007333
2016-03-19    0.015823
2016-03-20    0.020636
2016-03-21    0.020616
2016-03-22    0.021380
2016-03-23    0.018591
2016-03-24    0.019748
2016-03-25    0.019190
2016-03-26    0.016815
2016-03-27    0.015596
2016-03-28    0.020884
2016-03-29    0.022351
2016-03-30    0.024747
2016-03-31    0.023838
2016-04-01    0.022867
2016-04-02    0.024871
2016-04-03    0.025181
2016-04-04    0.024499
2016-04-05    0.124974
2016-04-06    0.221607
2016-04-07    0.132018
Name: last_seen, dtype: float64

`last_seen` range from 05-Mar-2016 to 07-Apr-2016 frequency percentage is less than 22%

### Dealing with Incorrect Registeration Year Data ###

In [175]:
autos['registeration_year'].describe()

count    48410.000000
mean      2004.773063
std         88.782344
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registeration_year, dtype: float64

`registeration_year` ranges from 1000 to 9999 years, so we must drop the invalid years. 
- We will remove the years below 1900, because automobiles were not invented back then. And we will remove years above 2016, because the car can not be registered after the listing was seen, so the data is inaccurate

In [176]:
#Count the number of listings with cars that fall outside the 1900-2016 interval
year_below_1900 = autos['registeration_year'] < 1900
len(year_below_1900)

48410

There are 48,410 data entries with years below 1900

In [177]:
autos.drop(autos[year_below_1900].index, inplace=True)

In [178]:
autos['registeration_year'].min()

1910

In [179]:
year_above_2016 = autos['registeration_year'] > 2016
len(year_above_2016)

48405

There are 48,405 data entries with years above 2016 

In [180]:
autos.drop(autos[year_above_2016].index, inplace=True)

In [181]:
autos['registeration_year'].max()

2016

In [183]:
#Calculate the percentage distribution of registeration year 
autos['registeration_year'].value_counts(normalize=True, dropna= False).sort_index()

1910    0.000086
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028152
2013    0.017235
2014    0.014226
2015    0.008381
2016    0.025982
Name: registeration_year, Length: 78, dtype: float64

`registeration_year` range from 1910 to 2016, the frequency percentage is no more than 3%

### Explore Car Brands ###
User aggregation using loops to explore variations across different car brands

In [203]:
brands = autos['brand'].value_counts(normalize=True)
brands, len(brands)

(volkswagen        0.211355
 bmw               0.109965
 opel              0.107322
 mercedes_benz     0.096448
 audi              0.086605
 ford              0.069993
 renault           0.047128
 peugeot           0.029871
 fiat              0.025681
 seat              0.018245
 skoda             0.016418
 nissan            0.015301
 mazda             0.015236
 smart             0.014205
 citroen           0.014033
 toyota            0.012744
 hyundai           0.010036
 sonstige_autos    0.009692
 volvo             0.009133
 mini              0.008789
 mitsubishi        0.008188
 honda             0.007844
 kia               0.007092
 alfa_romeo        0.006640
 porsche           0.006103
 suzuki            0.005931
 chevrolet         0.005695
 chrysler          0.003524
 dacia             0.002643
 daihatsu          0.002514
 jeep              0.002278
 subaru            0.002128
 land_rover        0.002106
 saab              0.001655
 jaguar            0.001569
 daewoo            0

In [212]:
top_brands = brands[brands > 0.05].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

Out of the 40 brands, the top 5% of the brands are volkswagen, bmw, opel, mercedes benz, audi and ford. We will use these brands to aggregate on. This is top 5% of the total counts. 

In [222]:
#Finding average price of the top brands 
brand_mean_prices = {}

for b in top_brands:
    #select all rows of that brand
    selected_rows = autos[autos['brand'] == b]
    #find the avg price of the selected row
    avg_price = int(selected_rows['price'].mean())
    #use key, value pair in dictionary 
    brand_mean_prices[b] = avg_price

brand_mean_prices

{'volkswagen': 5620,
 'bmw': 8606,
 'opel': 2991,
 'mercedes_benz': 8657,
 'audi': 9362,
 'ford': 4063}

- Opel and Ford are less expensive
- Audi, Mercedes Benz, BMW has are more expensive.
- Volkwagen is in between

In [237]:
#Finding the mileage of the top brands 
brand_mean_mileage = {}

for b in top_brands:
    #select all rows of that brand
    selected_rows = autos[autos['brand'] == b]
    #find avg mileage from selected rows 
    avg_mileage = int(selected_rows['odometer_km'].mean())
    #use key, value pair for the dictionary 
    brand_mean_mileage[b] = avg_mileage

brand_mean_mileage

{'volkswagen': 128710,
 'bmw': 132557,
 'opel': 129314,
 'mercedes_benz': 130838,
 'audi': 129208,
 'ford': 124210}

- BMW, Mercedes Benz has the most mileage.
- Volkswagen, Ford, Audi has the least mileage. 
- Opel is in between 

In [245]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending=False)
bmp_series

audi             9362
mercedes_benz    8657
bmw              8606
volkswagen       5620
ford             4063
opel             2991
dtype: int64

In [246]:
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
bmm_series

bmw              132557
mercedes_benz    130838
opel             129314
audi             129208
volkswagen       128710
ford             124210
dtype: int64

In [247]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
audi                 9362
mercedes_benz        8657
bmw                  8606
volkswagen           5620
ford                 4063
opel                 2991

In [249]:
df['mean_mileage'] = pd.Series(brand_mean_mileage)
df

mean_price  mean_mileage
audi                 9362        129208
mercedes_benz        8657        130838
bmw                  8606        132557
volkswagen           5620        128710
ford                 4063        124210
opel                 2991        129314

There seems to be no relations between price and mileage. For example, Audi, which is the most expensive, does not have the highest mileage. 
And Opel, which is the least expensive, does not have the lowest mileage. 